C++ API parity: ReLU

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27309 C++ API parity: MultiheadAttention
* #27382 C++ API parity: Linear
* #27538 C++ API parity: Threshold
* #27537 C++ API parity: Tanhshrink
* #27536 C++ API parity: Tanh
* #27535 C++ API parity: Softsign
* #27534 C++ API parity: Softshrink
* #27489 C++ API parity: Softplus
* #27488 C++ API parity: Sigmoid
* #27487 C++ API parity: CELU
* #27437 C++ API parity: RReLU
* #27436 C++ API parity: ReLU6
* **#27435 C++ API parity: ReLU**

Differential Revision: [D17835407](https://our.internmc.facebook.com/intern/diff/D17835407)

pbelevich committed:
C++ API parity: ReLU

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU"




[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/modules/activation.h:
```diff
@@ -146,5 +146,28 @@ class TORCH_API PReLUImpl : public torch::nn::Cloneable<PReLUImpl> {
+class TORCH_API ReLUImpl : public torch::nn::Cloneable<ReLUImpl> {
+ public:
+  ReLUImpl() : ReLUImpl(ReLUOptions()) {}
+  explicit ReLUImpl(const ReLUOptions& options_);
```
similarly here: we might be able to combine these two constructors.

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -116,5 +116,23 @@ void PReLUImpl::pretty_print(std::ostream& stream) const {
+void ReLUImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::ReLU(";
+  if (options.inplace()) {
+    stream << std::boolalpha  << "inplace=" << options.inplace();
```
nit:
```suggestion
    stream << std::boolalpha << "inplace=" << options.inplace();
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -44,6 +44,14 @@ inline Tensor prelu(const Tensor& input, const Tensor& weight) {
   return torch::prelu(input, weight);
 }
 
+inline Tensor relu(Tensor& input, const ReLUOptions& options) {
```
I think `options` can be optional in Python version, and it would be great to have a test for `relu(input)` case as well :)

pbelevich committed:
Update on "C++ API parity: ReLU"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU"




[ghstack-poisoned]

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -1420,3 +1437,9 @@ TEST_F(ModulesTest, PrettyPrintPReLU) {
+TEST_F(ModulesTest, PrettyPrintReLU) {
+  ASSERT_EQ(c10::str(ReLU()), "torch::nn::ReLU()");
+  ASSERT_EQ(c10::str(ReLU(ReLUOptions().inplace(true))),
+    "torch::nn::ReLU(inplace=true)");
```
It would be great to add a test case for `ReLU(/*inplace=*/true)` as well :D

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -380,3 +380,32 @@ TEST_F(FunctionalTest, PReLU) {
+    auto x = torch::linspace(-10.0, 10.0, size * size * size);
+    x.resize_({size, size, size});
+    auto y_exp = (x < 0) * 0 + (x >= 0) * x;
+    auto y = F::relu(x, ReLUOptions().inplace(inplace));
```
It would be great to add a test case for `F::relu(x, /*inplace=*/inplace)` as well :D

pbelevich committed:
Update on "C++ API parity: ReLU"


Differential Revision: [D17835407](https://our.internmc.facebook.com/intern/diff/D17835407)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU"


Differential Revision: [D17835407](https://our.internmc.facebook.com/intern/diff/D17835407)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU"


Differential Revision: [D17835407](https://our.internmc.facebook.com/intern/diff/D17835407)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU"


Differential Revision: [D17835407](https://our.internmc.facebook.com/intern/diff/D17835407)

[ghstack-poisoned]

yf225 reviewed (approved):
Thanks a lot @pbelevich !

pbelevich committed:
Update on "C++ API parity: ReLU"


Differential Revision: [D17835407](https://our.internmc.facebook.com/intern/diff/D17835407)

[ghstack-poisoned]

bddppq commented:
I think this commit is the culprit of the cpp tests failure on current master:
https://circleci.com/api/v1.1/project/github/pytorch/pytorch/3154023/output/105/0?file=true

```
Oct 10 22:30:06 ======================================================================
Oct 10 22:30:06 ERROR: test_cpp (__main__.TestJit)
Oct 10 22:30:06 ----------------------------------------------------------------------
Oct 10 22:30:06 Traceback (most recent call last):
Oct 10 22:30:06   File "/var/lib/jenkins/workspace/test/common_utils.py", line 206, in wrapper
Oct 10 22:30:06     fn(*args, **kwargs)
Oct 10 22:30:06   File "test_jit.py", line 2049, in test_cpp
Oct 10 22:30:06     torch._C._jit_run_cpp_tests(run_cuda=False)
Oct 10 22:30:06 RuntimeError: basic_string::_M_construct null not valid
Oct 10 22:30:06 
```
